pkg/model/model.go

```diff
@@ -0,0 +1,295 @@
+package model
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/h2non/filetype"
+	"github.com/h2non/filetype/matchers"
+	"github.com/h2non/filetype/types"
+	svg "github.com/h2non/go-is-svg"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+func init() {
+	t := types.NewType("svg", "image/svg+xml")
+	filetype.AddMatcher(t, svg.Is)
+	matchers.Image[types.NewType("svg", "image/svg+xml")] = svg.Is
+}
+
+type Model map[string]*Package
+
+func (m Model) Validate() error {
+	var result *multierror.Error
+
+	for name, pkg := range m {
+		if name != pkg.Name {
+			result = multierror.Append(result, fmt.Errorf("package key %q does not match package name %q", name, pkg.Name))
+		}
+		if err := pkg.Validate(); err != nil {
+			result = multierror.Append(result, fmt.Errorf("invalid package %q: %v", pkg.Name, err))
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+type Package struct {
+	Name           string
+	Description    string
+	Icon           *Icon
+	DefaultChannel *Channel
+	Channels       map[string]*Channel
+}
+
+func (m *Package) Validate() error {
+	var result *multierror.Error
+	if m.Name == "" {
+		result = multierror.Append(result, errors.New("package name must not be empty"))
+	}
+
+	if err := m.Icon.Validate(); err != nil {
+		result = multierror.Append(result, fmt.Errorf("invalid icon: %v", err))
+	}
+
+	if len(m.Channels) == 0 {
+		result = multierror.Append(result, fmt.Errorf("package must contain at least one channel"))
+	}
+
+	if m.DefaultChannel == nil {
+		result = multierror.Append(result, fmt.Errorf("default channel must be set"))
+	}
+
+	foundDefault := false
+	for name, ch := range m.Channels {
+		if name != ch.Name {
+			result = multierror.Append(result, fmt.Errorf("channel key %q does not match channel name %q", name, ch.Name))
+		}
+		if err := ch.Validate(); err != nil {
+			result = multierror.Append(result, fmt.Errorf("invalid channel %q: %v", ch.Name, err))
+		}
+		if ch == m.DefaultChannel {
+			foundDefault = true
+		}
+		if ch.Package != m {
+			result = multierror.Append(result, fmt.Errorf("channel %q not correctly linked to parent package", ch.Name))
+		}
+	}
+
+	if m.DefaultChannel != nil && !foundDefault {
+		result = multierror.Append(result, fmt.Errorf("default channel %q not found in channels list", m.DefaultChannel.Name))
+	}
+	return result.ErrorOrNil()
+}
+
+type Icon struct {
+	Data      []byte
+	MediaType string
+}
+
+func (i *Icon) Validate() error {
```

Comment:
The benefit of putting the validation on the model is that any representation that can be converted to the model can benefit from the exact same validation.

We could extract these somewhat peripheral validations into separate functions on the model that aren't always called when we're validating the core invariants.

Reply:
Simple validation to make sure an icon is actually an image and has the expected media type seems fine, otherwise the icon could be filled with junk that would not work with downstream consumers. This validation will always run on a package, unlike a validation command which may or may not have been run by the package maintainer.